### PR TITLE
simplify special badge code

### PR DESCRIPTION
**Pull request by johnduffell**
this follows on from @TBonnin comments on the special badging yesterday, now we always return a Badge to the templates, and only use SpecialBadge when we are working out whether it even applies.

cc @blongden73

### CODE REVIEW

File: common/app/model/Badges.scala [hunk 1]
```diff
@@ -8,19 +8,17 @@ import java.math.BigInteger
...
+  def maybeThisBadge(tag: String): Option[Badge]
 }
 
-case class Badge(seriesTag: String, staticImageUrl: String, classModifier: Option[String] = None) extends BaseBadge {
-  def isBadgeTag(tag: String) = seriesTag == tag
-  def imageUrl(id: String) = staticImageUrl
+case class Badge(seriesTag: String, imageUrl: String, classModifier: Option[String] = None) extends BaseBadge {
+  def maybeThisBadge(tag: String) = if (seriesTag == tag) Some(this) else None
 }
 case class SpecialBadge(hashedTag: String, classModifier: Option[String] = None) extends BaseBadge {
```
Comment (TBonnin) [inline common/app/model/Badges.scala]: do you need the `classModifier` param in `SpecialBadge`?

File: common/app/model/Badges.scala [hunk 2]
```diff
@@ -8,19 +8,17 @@ import java.math.BigInteger
...
-trait BaseBadge{
-  def imageUrl(id: String): String
-  def classModifier: Option[String]
-  def isBadgeTag(tag: String):Boolean
+trait BaseBadge {
+  def maybeThisBadge(tag: String): Option[Badge]
 }
 
-case class Badge(seriesTag: String, staticImageUrl: String, classModifier: Option[String] = None) extends BaseBadge {
-  def isBadgeTag(tag: String) = seriesTag == tag
```
Comment (TBonnin) [inline common/app/model/Badges.scala]: why have you not keep this `isBadgeForTag()` method? 
I found having a class method return `this` kind of weird. Also it mean using `flatmap` below when `find` is imho a better way to convey what the code is doing.

File: common/app/model/Badges.scala (hunk 1 above)
Comment (johnduffell) [inline common/app/model/Badges.scala]: good point, removed it

File: common/app/model/Badges.scala [hunk 3]
```diff
@@ -53,10 +53,20 @@ object Badges {
...
-  def badgeFor(c: ContentType) = c.tags.tags.map(_.id).foldLeft(None: Option[Badge]) { (maybeBadge, tag) =>
-      maybeBadge orElse allBadges.flatMap(b => b.maybeThisBadge(tag)).headOption
+  def badgeFor(c: ContentType) = {
+    badgeForTags(c.tags.tags.map(_.id))
   }
 
-  def badgeFor(fc: FaciaContainer) = fc.href.flatMap(href => allBadges.flatMap(b => b.maybeThisBadge(href)).headOption)
+  def badgeForTags(tags: Traversable[String]) = {
+
+    val badgesForTags = tags.flatMap {
```
Comment (TBonnin) [inline common/app/model/Badges.scala]: You didn't like the for comprehension? :wink:

File: common/app/model/Badges.scala [hunk 4]
```diff
@@ -53,10 +53,20 @@ object Badges {
...
+      case tagId => allBadges.flatMap {
+        case baseBadge => baseBadge.maybeThisBadge(tagId)
```
Comment (TBonnin) [inline common/app/model/Badges.scala]: Do you need cases here?
Also in most part of the code map/flatmap are formatted like below:
```
something.map { wrapped =>
    doingSomethingWith(wrapped)
    ...
}
```
Can we do the same here?

File: common/app/model/Badges.scala (hunk 3 above)
Comment (johnduffell) [inline common/app/model/Badges.scala]: oh I thought you didn't, and suggested I used flatmap =P

File: common/app/model/Badges.scala (hunk 4 above)
Comment (johnduffell) [inline common/app/model/Badges.scala]: true, I can revert to for comprehension anyway

**Comment by TBonnin** [comment]
👍

**Pull request merged**